### ISSUE TITLE
Server push in HTTP/2

### ISSUE THREAD

**Issue by kazu-yamamoto** [yesodweb/wai#415]
- WAI API does not allow to support server pushing which can be controlled by WAI applications.
- Jetty has a smart push mechanism with a learning dictionary based on Referer:. E.g. `default.css` is referred by `index.html`. So, if `index.html` is required next time, we can push `default.css`. It's not a good idea to support this by Warp itself. Rather, we should provide a library which WAI applications can use.
- Another idea is cookie. If a client tells which resources are necessary via a cookie, a server can surely push them.

**Comment by awpr** [comment]
One observation: if the separate Http2Application type is accepted, the normal WAI Application will not grow an explicit interface for pushing streams, as Warp's HTTP/2 stream-push handling will be explicit in the Http2Application type.  In this case any of the several proposed "push inference" mechanisms would necessarily take the form of HTTP/2-only middleware.  This would mean it's applicable to natively-HTTP/2 applications as well as normal Applications served over HTTP/2 (by composing it with promoteApplication), but not to pure-HTTP/1 applications.  This seems to be the desirable outcome anyway (since it's nonsense for an application being served over HTTP/1 to attempt to push a stream).

Unfortunately it would require callers to use 'runHttp2 port (smartPush $ promoteApplication myApp) myApp' instead of simply 'run port myApp'.  Maybe Settings could grow an 'upgradeware' field for middleware that's automatically applied to promoted apps, to allow 'runSettings (setPort port $ setUpgradeware smartPush defaultSettings) myApp' instead.

**Comment by davean** [comment]
I generally prefer there be a different interface for HTTP2. I for one can
make far better use of it by custom coding to it then any middleware could
hope to.

It would be easy to provide HTTP and HTTP2 versions for me, and a HTTP to
HTTP2 wrapper appears easy for the cases that don't write custom code. This
approach also trivially allows HTTP2 middleware.

**Comment by awpr** [comment]
@davean We're in agreement -- see https://github.com/awpr/wai/commits/wai_http2 and https://github.com/yesodweb/wai/issues/402.

My comment here was only concerning the interface of an automatic server push middleware that would be implemented against the explicit interface (which would be available only to HTTP/2-aware applications).

**Comment by awpr** [comment]
Initial implementation is now pushed to https://github.com/awpr/wai/commits/wai_http2.  Some of the types are still a bit of a mess (e.g. Applications have to use the Zoidberg operator to make a Response, the Request in HTTP2 is an alias for (H1.Request, PushFn)), but the mechanics are there, and Chrome consistently accepts the streams except when the application pushes two streams of the same URL simultaneously, which for some reason it treats as a protocol error.  Maybe that's just leftover logic from SPDY.  I'll file a bug report for that soon.

**Comment by AaronFriel** [comment]
@awpr Do you have a design doc or an issue somewhere where you're tracking the API changes and discussing rationale? I haven't had a chance to dig into the commits, but it would be great to see something like a Rust PR for this. Admittedly, I don't understand HTTP/2.0 as well as I ought to, and I'd guess it pretty educational to understand why someone at Google is doing things a certain way with it.

**Comment by kazu-yamamoto** [comment]
@AaronFriel This does not directly answer to your question, just FYI. I'm planning to write some blog articles about design and implementation of Warp HTTP/2.

But Daniel has come to Japan as IIJ-II (our company) intern. We should concentrate to develop ghc-mod for the next two months. So, I guess I can write them in October.

**Comment by awpr** [comment]
@AaronFriel the discussion on https://github.com/yesodweb/wai/issues/402 is the best I've currently got for design rationale; I could probably write a more detailed explanation of decisions once it's more final.

**Comment by awpr** [comment]
Update on this: push support is mostly cleaned up; there are 4 things I'd still like to fix.

- It doesn't currently honor the client's MAX_CONCURRENT_STREAMS setting.
- If the sender happens to encounter a stream's trailers (end of stream) before the push attempt, it will just ignore the push; it would be nice for PUSH_PROMISE frames to participate in the ordering with the associated stream's DATA frames, so it wouldn't be possible for a push to get ignored because of unfortunate ordering by the priority tree.
- API issue: most applications will probably want to push a stream as if it were a request to the application itself, providing just a request to serve (e.g. "push /style.css to the client"), but some applications might want to provide the response directly as a stream.  Currently only the latter is supported.
- Mechanics issue: currently 'pushPromise' just runs the streaming body fully and then returns; most applications will probably want to serve pushed streams concurrently.  I should probably just change it to 'forkIO'.

**Comment by kazu-yamamoto** [comment]
Server push in H2O: https://github.com/h2o/h2o/issues/421

**Comment by awpr** [comment]
I've started working on this again the last few days; another update:

* It now honors MAX_CONCURRENT_STREAMS and ENABLE_PUSH.
* The race issue is addressed by waiting for the sender to report back on a pushed stream before returning from pushResponder.
* No progress on pushing uses of the HTTP2Application itself.  It's not impossible, just unwieldy, and we should ideally make it more ... wieldy.
* It now runs the pushed stream in its own thread.  We can always revisit this if people want finer control.
* I added back support for sending file chunks (as discussed on #402).  It's mostly the same as the ResponseFile handling before.

I don't have any further compatibility-breaking changes to HTTP2Application in mind, and I think all the blocking issues have been addressed, so I think it's nearly time to send the pull request.  First I need to test the file chunk sending code on Windows, and I'll want to look over the changes for readability/comments/cleanup.
